sys/kern/proc.c

```diff
@@ -648,11 +683,21 @@ int do_waitpid(pid_t pid, int *status, int options, pid_t *cldpidp) {
         return error;
       }
 
-      /* Wait until one of children changes a state. */
-      klog("PID(%d) waits for children (pid = %d)", p->p_pid, pid);
-      cv_wait(&p->p_waitcv, all_proc_mtx);
+      proc_lock(p);
+      if (p->p_flags & PF_CHILD_STATE_CHANGED) {
+        /* A child changed state while we were checking: recheck. */
+        p->p_flags &= ~PF_CHILD_STATE_CHANGED;
+        proc_unlock(p);
+        goto check_children;
```

Comment:
Can we make here a normal loop without `goto`?

Reply:
We could, but a loop seems out of place here, since there's no actual looping that's happening: we would need to insert a `break` at the end of the loop, and `goto check_children` would be replaced by `continue`. I think `goto` is the better choice in this case.

Reply:
Okay.